src/python/pants/engine/query.py

```diff
@@ -0,0 +1,239 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import ast
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Dict, Tuple, Type, TypeVar
+
+from pants.build_graph.address import Address
+from pants.engine.addresses import Addresses
+from pants.engine.internals.graph import Owners, OwnersRequest
+from pants.engine.rules import RootRule, rule
+from pants.engine.selectors import Get
+from pants.engine.unions import UnionMembership, UnionRule, union
+from pants.scm.subsystems.changed import ChangedOptions, ChangedAddresses, ChangedRequest, DependeesOption, UncachedScmWrapper
+from pants.util.meta import classproperty
+from pants.util.strutil import safe_shlex_split
+
+
+@union
+class QueryComponent(ABC):
+
+    @classproperty
+    @abstractmethod
+    def function_name(cls):
+        """The initial argument of a shlexed query expression.
+
+        If the user provides --query='<name> <args...>' on the command line, and `<name>` matches this
+        property, the .parse_from_args() method is invoked with `<args...>` (shlexed, so split by
+        spaces).
+        """
+
+    @classmethod
+    @abstractmethod
+    def parse_from_args(cls, *args):
+        """Create an instance of this class from variadic positional string arguments.
+
+        This method should raise an error if the args are incorrect or invalid.
+        """
+
+
+@dataclass(frozen=True)
+class QueryAddresses:
+    addresses: Addresses
```

Comment:
More idiomatic is to use `Collection[Address`]:

https://github.com/pantsbuild/pants/blob/22cc05f53fe5e4619e42f6ada5d13efef1a2ec59/src/python/pants/engine/internals/graph.py#L187-L188